### DISCUSSION TITLE
SNMP is broken?

### DISCUSSION THREAD

**Discussion by greatest-cornholio** [jgyates/genmon#1165] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).
- [X] I have included the generator make and model in my submission.

### Version

Raspbian and Raspberry Pi OS from buster-based to current version, both desktop and lite, both 32 and 64 bit. Genmon 1.19.03

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have not submitted logs

### What is/are your question(s)?

Hey all
We have a couple Pi3 B devices running genmon software. I've been tasked with building more. The two that we have up and running are on Raspbian Buster, with genmon fully updated. The issue I'm running into is that I can't make SNMP work. I've tried numerous versions of Raspberry Pi OS and Raspbian going back to debian 10 based versions, with no luck. I've tried both 64 and 32 bit versions, as well. I managed to SNMP walk one running the newest version, but only after installing a bunch of additional packages via APT, fixing a bunch of permissions, and manually starting gensnmp.py with sudo. I've since wiped the card that instance was on, and didn't really document what I did to make it work. 

When I run ./startgenmonmaint.sh stop, I always get a "No such process" error directly after "stopping genmon.py"

Is there something obvious I'm missing? I've searched high and low, and apparently nobody else is having this issue, but any input any of you have... I'll try it.

FWIW, these will be monitoring Generac Nexus-based devices, but until I get SNMP to work, I'm just bench testing them on my desk. When I do the full install and take them to the site, everything works as expected except that we can't SNMPwalk them, and our LibreNMS instance can't see them.

Thanks!

**Comment by jgyates** [comment]
SNMP is working for me on both 32 and 64 bit platforms. If you want to send your logs I can take a look at them.

How are you installing genmon? Are you following the instructions here: https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software

Regarding this statement: 

"When I run ./startgenmonmaint.sh stop, I always get a "No such process" error directly after "stopping genmon.py"

This would indicate that genmon is not loaded. If you are having and issue loading genmon, then the SNMP add on will not work.

What is the output of /var/log/genmon.log and /var/log/gensnmp.log?

**Comment by greatest-cornholio** [comment]
I attempted this again and took detailed notes (documentation.txt, attached). In a nutshell, I did install following the instructions provided. The genmon process itself does work, and work as expected. Meaning if I take the current build on my desk and test it at our site, it would do everything we need it to... Except communicate via snmp. I mistyped what I said previously... the "No such process" comes directly after "stopping gensnmp.py." 

When I try to run 'python3 gensnmp.py', I get the following:
Error loading pysnmp! :No module named 'pyasn1'

In addition to my notes, I'm attaching genmon.log, as requested. There is no gensnmp.log file. I ran the script with a tee command to dump the output of the install script into a log so I'll include that as well.

Thanks for this awesome project and community!

[Documentation.txt](https://github.com/user-attachments/files/16617680/Documentation.txt)
[genmon.log](https://github.com/user-attachments/files/16617681/genmon.log)
[genmoninstall.log](https://github.com/user-attachments/files/16617682/genmoninstall.log)

**Reply by jgyates** [reply]
According to your logs you posted you are running in a managed python environment (this is the default on the Raspberry pi OS with the current version of the OS). Genmon, and the required libraries will be installed in it's own virtual python environment as required on a managed python environment. More info is here: https://github.com/jgyates/genmon/wiki/Appendix-S---Working-in-a-Managed-Python-Environment

The failure you are getting when attempting to load gensnmp.py is due to you trying to load this add on program outside of the virtual environment that the required libraries are installed into.

The python script startgenmon.sh handles the enabling the python environment so I would start and stop genmon with the provided script rather than starting it with the "python gensnmp.py" command. 

If you did want to start gensnmp.py out side of the script then you would need to type this:

     cd genmon
     source ./genenv/bin/activate
     sudo ./genenv/bin/python3 ./gensnmmp.py

To deactivate the virtual environment type:

     deactivate

If you use the above method the program will remain loaded until you reboot or kill it manually with the kill or pskill command.

The script startgenmon.sh will load the virtual environment and use genloader.py to start and stop genmon and any add on programs. genloader.py tracks the process ID of genmon and any loaded add on programs so it can kill / stop them later.

The way genmon was designed to be used is to install with the script, then attached it to your generator, load the web interface and select the add on programs you want to load from the web interface.  If starting or stoping genmon and the add on programs, use startgenmon.sh. 

If you are trying to load genmon and the add on programs without being connected to a generator then I would suggest not using that scenario as this does not allow genmon to identify the controller and it will never complete the initialization process.

**Reply by jgyates** [reply]
If you enabled the SNMP add on via the web interface, do you have a file /var/log/gensnmp.log? If yes then what is in this file?

**Reply by jgyates** [reply]
FYI, I just tried loading genmon with the snmp add on with no controller connected and it does load and respond with the default controller (Nexus Air cooled). you may try loading genmon with the script and then manually load gensnmp.py with the method above. It is possible that there is a dependency problem with some of the required libraries. Are you using hte 32 bot or 64 bit version of the OS?

**Reply by greatest-cornholio** [reply]
I can certainly run up to our network site and attempt some more diagnostics with the device connected to a controller, but we were getting similar issues with it connected earlier last week, so I brought it back to my desk to bench test. However, if we get better data connected, I'll try to get up there tomorrow.

I've tried multiple operating systems on the Pi, but currently, I'm running the newest 64 bit Raspberry Pi OS Lite (Debian 12 - Bookworm based). I get pretty much the same thing as far back as Buster-based Raspbian and / or Raspbian Lite. Haven't gone back further than that.

I did enable SNMP via the genmon web interface, but I do not have a gensnmp.log file in /var/log/. 

I attempted to launch gensnmp from within genenv as instructed, (had to change it slightly to 'sudo ./genenv/bin/python3 addon/gensnmp.py' because the command above attempted to run gensnmp.py from ~/genmon/./ instead of ~/genmon/addon/, and couldn't find the file)
When I attempt to launch gensnmp from the environment, I get this error:
Error loading pysnmp! :No module named 'pysnmp.carrier.asyncore'

A google search turned up this thread that looked potentially related, so I did what was linked at https://github.com/home-assistant/core/issues/110100#issuecomment-1961023901 from within genenv, and it didn't help. Hopefully it didn't hurt...

**Reply by jgyates** [reply, marked as answer]
I did some testing and updated to the latest version of pysnmp and I think I found the issue. With pysnmp version 6.x it appears they have changed some of the internal locations of functions. I updated the genmon requirements file to only use pysnmp version 5.1.0. You can update to the latest from the command prompt with these two commands:

first to update your github repo to get my latest change, from the genmon folder type:

      ./genmonmaint.sh -u

Then to refresh your libaries that genmon uses type:

     ./genmonmaint.sh -r

If you started any programs manually (i.e. not with startgenmon.sh then reboot) otherwise type this to restart genmon:

     ./startgenmon.sh restart

Let me know how it goes for you.

**Reply by greatest-cornholio** [reply]
Yup that did the trick!!! Thanks so much, this is a fantastic piece of software!!

**Reply by jgyates** [reply]
Thanks, I am glad it is working for you now.
